library/L1_Peripheral/lpc40xx/test/timer_test.cpp

```diff
@@ -4,5 +4,132 @@
+  When(Method(mock_system_controller, GetPeripheralClockDivider))
+      .AlwaysReturn(kExpectedPeripheralClockDivider);
+  sjsu::SystemController::SetPlatformController(&mock_system_controller.get());
+
+  const Timer::Peripheral_t kTimerPeripheral = {
+    .peripheral = &local_timer_registers,
+    .id         = kExpectedPeripheralId,
+    .irq        = kExpectedIrq,
+  };
+  Timer timer(kTimerPeripheral);
+
+  SECTION("Initialize")
+  {
+    constexpr units::frequency::hertz_t kExpectedFrequency = 1_MHz;
+    constexpr uint32_t kExpectedPrescaler =
+        kSystemControllerClockFrequency / kExpectedFrequency;
+    const Status kStatus = timer.Initialize(kExpectedFrequency);
+
+    Verify(Method(mock_system_controller, PowerUpPeripheral)
+               .Matching([](sjsu::SystemController::PeripheralID id) {
+                 return kExpectedPeripheralId.device_id == id.device_id;
+               }),
+           Method(mock_interrupt_controller, Enable));
+    CHECK(kStatus == Status::kSuccess);
+    CHECK(local_timer_registers.PR == kExpectedPrescaler);
+  }
+
+  SECTION("SetMatchBehavior")
+  {
+    constexpr std::array<uint32_t, kExpectedMatchRegisterCount> kTicks = {
+      1'000, 100'000, 100, 123
+    };
+    constexpr std::array kMatchActions = {
+      Timer::MatchAction::kInterrupt,
+      Timer::MatchAction::kRestart,
+      Timer::MatchAction::kStop,
+      Timer::MatchAction::kInterruptRestart,
+      Timer::MatchAction::kInterruptStop,
+      Timer::MatchAction::kRestartStop,
+      Timer::MatchAction::kInterruptRestartStop
+    };
+    volatile uint32_t * match_register_pointer = &local_timer_registers.MR0;
+
+    for (size_t i = 0; i < kExpectedMatchRegisterCount; i++)
+    {
+      const uint8_t kExpectedMatchRegister = i;
+      for (size_t j = 0; j < kMatchActions.size(); j++)
+      {
+        const Timer::MatchAction kMatchAction = kMatchActions[j];
+        const bit::Mask kMatchActionMask      = {
+          .position = static_cast<uint8_t>(kExpectedMatchRegister * 3),
+          .width    = 3,
+        };
+
+        INFO("ticks: " << static_cast<size_t>(kTicks[i]));
+        INFO("action: " << static_cast<size_t>(kMatchActions[j]));
+        INFO("match_register: " << i);
+
+        timer.SetMatchBehavior(kTicks[i], kMatchActions[j],
+                               kExpectedMatchRegister);
+
+        CHECK(bit::Extract(local_timer_registers.MCR, kMatchActionMask) ==
+              static_cast<uint8_t>(kMatchAction));
+        CHECK(match_register_pointer[kExpectedMatchRegister] == kTicks[i]);
+      }
+    }
+  }
+
+  SECTION("GetAvailableMatchRegisters")
+  {
+    CHECK(timer.GetAvailableMatchRegisters() == kExpectedMatchRegisterCount);
+  }
+
+  SECTION("GetCount")
+  {
+    CHECK(local_timer_registers.TC == 0);
+    CHECK(timer.GetCount() == 0);
+
+    constexpr uint32_t kExpectedCount = 12'345;
+    local_timer_registers.TC          = kExpectedCount;
+    CHECK(timer.GetCount() == kExpectedCount);
```

Comment:
```suggestion
    // Setup
    constexpr uint32_t kExpectedCount = 12'345;
    local_timer_registers.TC          = kExpectedCount;
    
    // Exercise
    uint32_t actual_count = timer.GetCount();
    
    // Verify
    CHECK(actual_count == kExpectedCount);
```

Can you format your tests with a "setup", "exercise" and "verify" regions with comments. This is something that I learned about at work. And it really helps to organize your test cases.

I would also make another test sections just for testing that GetCount() defaults to 0 when TC is zero, rather than overload it in this test section. Each test section should have 1 setup, exercise, verify. But you can loop the exercise and verify as long as the thing you are testing does not change subjects.

Reply:
Make sure to do this for the other sections as well.